### PR TITLE
feat(entities): added entity capabilities service

### PR DESCRIPTION
**Pull request by jdalsem**

### CODE REVIEW

File: engine/classes/Elgg/EntityCapabilitiesService.php [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	
+	/**
+	 * Stores the entities with their capabilities
+	 *
+	 * @var array
+	 */
+	protected $entities = [];
+	
+	/**
+	 * Checks if an entity has a specific capability
```
Comment (jeabakker) [inline engine/classes/Elgg/EntityCapabilitiesService.php:21]: it's not an entity, but an type/subtype combination

File: engine/classes/Elgg/EntityCapabilitiesService.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	 *
+	 * @param string $type       type of the entity
+	 * @param string $subtype    subtype of the entity
+	 * @param string $capability name of the capability to check
+	 * @param bool   $default    default value to return if it is not explicitly set
+	 *
+	 * @return bool
+	 */
+	public function hasCapability(string $type, string $subtype, string $capability, bool $default = false): bool {
+		return elgg_extract($capability, elgg_extract($subtype, $this->entities[$type], []), $default);
```
Comment (jeabakker) [inline engine/classes/Elgg/EntityCapabilitiesService.php:31]: `$this->entities[$type]` will cause a PHP warning if `$type` isn't set in the array

File: engine/classes/Elgg/EntityCapabilitiesService.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	 *
+	 * @param string $type       type of the entity
+	 * @param string $subtype    subtype of the entity
+	 * @param string $capability name of the capability to set
+	 *
+	 * @return void
+	 */
+	public function disableCapability(string $type, string $subtype, string $capability): void {
+		$this->setCapability($type, $subtype, $capability, false);
+	}
```
Comment (jeabakker) [inline engine/classes/Elgg/EntityCapabilitiesService.php:80]: are these needed? why not call `setCapability()` directly

File: engine/classes/ElggEntity.php [hunk 4]
```diff
@@ -2081,4 +2081,16 @@ public function invalidateCache() {
 			_elgg_services()->dataCache->get($namespace)->delete($this->guid);
 		}
 	}
+	
+	/**
+	 * Checks if an entity has a specific capability
+	 *
+	 * @param string $capability
```
Comment (jeabakker) [inline engine/classes/ElggEntity.php:2088]: param docs

File: engine/lib/entities.php [hunk 5]
```diff
@@ -715,142 +715,56 @@ function get_entity_statistics(int $owner_guid = 0): array {
...
-			return false;
-		}
-	} else {
-		unset($entities[$type]);
-	}
-
-	_elgg_services()->config->registered_entities = $entities;
-	return true;
+function elgg_entity_enable_capability(string $type, string $subtype, string $capability): void {
+	_elgg_services()->entity_capabilities->enableCapability($type, $subtype, $capability);
```
Comment (jeabakker) [inline engine/lib/entities.php:743]: simply use `setCapability()`

File: mod/externalpages/elgg-plugin.php [hunk 6]
```diff
@@ -8,17 +8,23 @@
 		[
 			'type' => 'object',
 			'subtype' => 'about',
-			'searchable' => false,
+			'capabilities' => [
+				'searchable' => true,
```
Comment (jeabakker) [inline mod/externalpages/elgg-plugin.php:12]: it was `false` why is it now `true`?

**Review by jeabakker — CHANGES_REQUESTED** [review]

File: engine/classes/Elgg/EntityCapabilitiesService.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+ * @since 4.1
+ */
+class EntityCapabilitiesService {
+	
+	/**
+	 * Stores the entities with their capabilities
+	 *
+	 * @var array
+	 */
+	protected $entities = [];
```
[the remark this answers is not in the document]
Comment (jdalsem) [reply engine/classes/Elgg/EntityCapabilitiesService.php:18]: ignored

**Review by jeabakker — APPROVED** [review]

**Pull request merged**